src/wing/CCPACSWingCell.cpp

```diff
@@ -663,33 +815,139 @@ void CCPACSWingCell::BuildSkinGeometry(GeometryCache& cache) const
     // get the shape of the skin
     TopoDS_Shape loftShape = m_parent->GetParent()->GetLoftSide() == UPPER_SIDE ? wsr.GetUpperShape() : wsr.GetLowerShape();
 
-    // calculate corner points on chord face and project points on loftShape
-    // cache points for later use
-    cache.IBLE = wsr.GetPoint(m_etaXsiCache->innerLeadingEdgePoint.eta,
-                              m_etaXsiCache->innerLeadingEdgePoint.xsi,
-                              WING_COORDINATE_SYSTEM);
-    cache.projectedIBLE = ProjectPointOnShape(loftShape, cache.IBLE, zRefDir);
-    cache.IBTE = wsr.GetPoint(m_etaXsiCache->innerTrailingEdgePoint.eta,
-                              m_etaXsiCache->innerTrailingEdgePoint.xsi,
-                              WING_COORDINATE_SYSTEM);
-    cache.projectedIBTE = ProjectPointOnShape(loftShape, cache.IBTE, zRefDir);
-    cache.OBLE = wsr.GetPoint(m_etaXsiCache->outerLeadingEdgePoint.eta,
-                              m_etaXsiCache->outerLeadingEdgePoint.xsi,
-                              WING_COORDINATE_SYSTEM);
-    cache.projectedOBLE = ProjectPointOnShape(loftShape, cache.OBLE, zRefDir);
-    cache.OBTE = wsr.GetPoint(m_etaXsiCache->outerTrailingEdgePoint.eta,
-                              m_etaXsiCache->outerTrailingEdgePoint.xsi,
-                              WING_COORDINATE_SYSTEM);
-    cache.projectedOBTE = ProjectPointOnShape(loftShape, cache.OBTE, zRefDir);
-
-    // cut the shape at the cell borders
-    TopoDS_Shape resultShape = CutSpanwise(cache, loftShape, SpanWiseBorder::Inner, m_positioningInnerBorder, zRefDir, 1e-2);
-    resultShape              = CutSpanwise(cache, resultShape, SpanWiseBorder::Outer, m_positioningOuterBorder, zRefDir, 1e-2);
-    resultShape              = CutChordwise(cache, resultShape, ChordWiseBorder::LE, m_positioningLeadingEdge, zRefDir, 1e-2);
-    resultShape              = CutChordwise(cache, resultShape, ChordWiseBorder::TE, m_positioningTrailingEdge, zRefDir, 1e-2);
-
-    // store the result
-    cache.skinGeometry = resultShape;
+    int use_contour_coordinates = (int)(m_positioningInnerBorder.GetInputType()  == CCPACSWingCellPositionSpanwise::Contour);
+    use_contour_coordinates    += (int)(m_positioningOuterBorder.GetInputType()  == CCPACSWingCellPositionSpanwise::Contour);
+    use_contour_coordinates    += (int)(m_positioningLeadingEdge.GetInputType()  == CCPACSWingCellPositionChordwise::Contour);
+    use_contour_coordinates    += (int)(m_positioningTrailingEdge.GetInputType() == CCPACSWingCellPositionChordwise::Contour);
+
+    if ( use_contour_coordinates > 0 && use_contour_coordinates != 4 ) {
+        throw CTiglError("If one boundary of a wing cell is defined using contour coordinates, "
+                         "all boundaries must be defined using contour coordinates");
+    }
+
+    if ( use_contour_coordinates ) {
```

Comment:
What, if you have mixed contour coordinates and corner point? Or does CPACS not allow this?

Reply:
No, I decided to exclude this case because it is complicated and probably not very robust: If one border is defined wrt chord face coordinates and another using uv-coordinates on the wing skin, we need to project either the chordface point onto the surface or vice versa to calculate the coordinates of the shared corner of the two borders. Also, this projection is expected to be inprecise near the leading edge, where the chordface and the wing skin are almost orthogonal to each other.

Finally, mixing border definitions is not an important use case in my opition.

Reply:
Maybe we should at least detect this case and warn, that this is not supported.

Reply:
Like this?

https://github.com/DLR-SC/tigl/blob/df6d8d273701bd625ade83b3c409238b62a5e035/src/wing/CCPACSWingCell.cpp#L823-L826

Or do you mean something different?

Reply:
Exactly!